File: stream-processing-with-pyflink/src/chapter6/co_process_function_timers.py
import argparse
import os
import json
import datetime
from typing import Tuple

from pyflink.common.typeinfo import Types
from pyflink.datastream import (
    DataStream,
    StreamExecutionEnvironment,
    RuntimeExecutionMode,
    TimeCharacteristic,
)
from pyflink.datastream.state import ValueStateDescriptor, ValueState
from pyflink.datastream.functions import CoProcessFunction, RuntimeContext
from pyflink.datastream.time_domain import TimeDomain
from pyflink.table import StreamTableEnvironment

from utils.model import SensorReading

###
### Incomplete as on_timer() method is not yet implemented
###


class ReadingFilter(CoProcessFunction):
    def __init__(self):
        self.forwarding_enabled = None
        self.disable_timer = None

    class OnTimerContext(CoProcessFunction.Context):
        def time_domain(self) -> TimeDomain:
            pass

    def open(self, runtime_context: RuntimeContext):
        self.forwarding_enabled = runtime_context.get_state(
            ValueStateDescriptor("filter_switch", Types.BOOLEAN())
        )
        self.disable_timer = runtime_context.get_state(ValueStateDescriptor("timer", Types.LONG()))

    def process_element1(self, value: SensorReading, ctx: CoProcessFunction.Context):
        if self.forwarding_enabled.value():
            yield value

    def process_element2(self, value: Tuple[str, int], ctx: CoProcessFunction.Context):
        # enable reading forwarding
        self.forwarding_enabled.update(True)
        # set disable forward timer
        cpt = ctx.timer_service().current_processing_time()
        timer_timestamp = cpt + value[1]
        cur_timer_timestamp = self.disable_timer.value() or 0
        print(f"{value} - {cpt} - {timer_timestamp} - {self.disable_timer.value()}")
        if timer_timestamp > cur_timer_timestamp:
            # remove current timer and register new timer
            ctx.timer_service().delete_processing_time_timer(cur_timer_timestamp)
            ctx.timer_service().register_processing_time_timer(timer_timestamp)
            self.disable_timer.update(timer_timestamp)

    # def on_timer(self, timestamp: int, ctx: OnTimerContext):
    #     print("TIMER CALLED")
    #     import ipdb

    #     ipdb.set_trace()
    #     self.forwarding_enabled.clear()
    #     self.disable_timer.clear()


def define_workflow(source_stream: DataStream, filter_switches: DataStream):
    pass


if __name__ == "__main__":
    """
    ## local execution
    python src/chapter5/process_function_timers.py
    """
    RUNTIME_ENV = os.getenv("RUNTIME_ENV", "local")

    parser = argparse.ArgumentParser()
    parser.add_argument("--verbose", action="store_true")
    parser.set_defaults(verbose=False)
    args = parser.parse_args()

    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_runtime_mode(RuntimeExecutionMode.STREAMING)
    env.set_stream_time_characteristic(TimeCharacteristic.ProcessingTime)
    if RUNTIME_ENV == "local":
        SRC_DIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
        jar_files = ["flink-faker-0.5.3.jar"]
        jar_paths = tuple([f"file://{os.path.join(SRC_DIR, 'jars', name)}" for name in jar_files])
        print(jar_paths)
        env.add_jars(*jar_paths)

    t_env = StreamTableEnvironment.create(stream_execution_environment=env)
    t_env.get_config().set_local_timezone("Australia/Sydney")
    t_env.execute_sql(
        """
        CREATE TABLE sensor_source (
            `id`        INT,
            `rn`        INT,
            `log_time`  TIMESTAMP_LTZ(3)
        )
        WITH (
            'connector' = 'faker',
            'rows-per-second' = '1',
            'fields.id.expression' = '#{number.numberBetween ''0'',''4''}',
            'fields.rn.expression' = '#{number.numberBetween ''0'',''100''}',
            'fields.log_time.expression' =  '#{date.past ''10'',''5'',''SECONDS''}'
        );
        """
    )

    source_stream = t_env.to_append_stream(
        t_env.from_path("sensor_source"),
        Types.TUPLE([Types.INT(), Types.INT(), Types.SQL_TIMESTAMP()]),
    )

    filter_switches = env.from_collection(collection=[("sensor_0", 5 * 1000)])

    forward_readings = (
        source_stream.map(SensorReading.from_tuple)
        .connect(filter_switches)
        .key_by(lambda s: s.id, lambda f: f[0])
        .process(ReadingFilter())
    )

    forward_readings.print()

    # source_stream.map(SensorReading.from_tuple).print()

    env.execute("Monitor sensor temperatures.")

File: stream-processing-with-flink/src/main/kotlin/me/jaehyeon/chapter6/CoProcessFunctionTimers.kt
package me.jaehyeon.chapter6

import me.jaehyeon.sensor.SensorReading
import me.jaehyeon.sensor.SensorSource
import org.apache.flink.api.common.eventtime.WatermarkStrategy
import org.apache.flink.api.common.state.ValueState
import org.apache.flink.api.common.state.ValueStateDescriptor
import org.apache.flink.api.common.typeinfo.Types
import org.apache.flink.api.connector.source.util.ratelimit.RateLimiterStrategy
import org.apache.flink.api.java.tuple.Tuple2
import org.apache.flink.configuration.Configuration
import org.apache.flink.connector.datagen.source.DataGeneratorSource
import org.apache.flink.streaming.api.datastream.DataStreamSource
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.apache.flink.streaming.api.functions.co.KeyedCoProcessFunction
import org.apache.flink.util.Collector

/**
 * This Flink job demonstrates a `KeyedCoProcessFunction` to implement a dynamically
 * configurable filter.
 *
 * The goal is to forward sensor readings for a specific sensor only for a limited
 * period of time, which is controlled by a second "control" stream.
 *
 * The pipeline works as follows:
 * 1. **Sensor Stream**: A high-volume stream of `SensorReading` events.
 * 2. **Control Stream**: A low-volume stream of `Tuple2<String, Long>` which represents
 *    filter commands: `(sensorId, durationInMillis)`.
 * 3. **KeyBy & Connect**: BOTH streams are keyed by the sensor ID. This is crucial as it
 *    ensures that a filter command for "sensor_2" is processed on the same physical
 *    task that handles readings for "sensor_2".
 * 4. **KeyedCoProcessFunction (`ReadingFilter`)**:
 *    - When a filter command arrives, it enables forwarding for that key and sets a
 *      processing time timer to disable it later.
 *    - When a sensor reading arrives, it checks the state for that key. If forwarding is
 *      enabled, the reading is passed through; otherwise, it is dropped.
 *    - When a timer fires, it disables forwarding for that key.
 */
object CoProcessFunctionTimers {
    @JvmStatic
    fun main(args: Array<String>) {
        val env = StreamExecutionEnvironment.getExecutionEnvironment()

        // Source for sensor readings (using processing time)
        val readings =
            env.fromSource(
                SensorSource(),
                WatermarkStrategy.noWatermarks(),
                "Sensor Source",
            )

        // Source for filter switch commands: (sensorId, duration)
        val filterSwitches = FilterSwitch.createFilterSwitchSource(env)

        // Key BOTH streams by sensor ID and then connect them
        val forwardedReadings =
            readings
                .keyBy { it.id }
                .connect(filterSwitches.keyBy { it.f0 })
                .process(ReadingFilter())

        forwardedReadings.print()

        env.execute("CoProcessFunction Timers Example")
    }
}

/**
 * A KeyedCoProcessFunction that forwards readings for a given key for a limited
 * time based on commands from a second stream.
 */
class ReadingFilter : KeyedCoProcessFunction<String, SensorReading, Tuple2<String, Long>, SensorReading>() {
    // State to track whether forwarding is currently enabled for this key
    private lateinit var forwardingEnabled: ValueState<Boolean>

    // State to track the timestamp of the disable timer for this key
    private lateinit var disableTimer: ValueState<Long>

    override fun open(parameters: Configuration) {
        // Initialize state descriptors
        val forwardingEnabledDescriptor = ValueStateDescriptor("forwardingEnabled", Types.BOOLEAN)
        forwardingEnabled = runtimeContext.getState(forwardingEnabledDescriptor)

        val disableTimerDescriptor = ValueStateDescriptor("disableTimer", Types.LONG)
        disableTimer = runtimeContext.getState(disableTimerDescriptor)
    }

    /** This method is called for each element from the FIRST stream (SensorReadings). */
    override fun processElement1(
        value: SensorReading,
        ctx: Context,
        out: Collector<SensorReading>,
    ) {
        // Check if forwarding is enabled for the current key (defaults to false if null)
        val isForwarding = forwardingEnabled.value() ?: false
        if (isForwarding) {
            out.collect(value)
        }
    }

    /** This method is called for each element from the SECOND stream (filter commands). */
    override fun processElement2(
        value: Tuple2<String, Long>,
        ctx: Context,
        out: Collector<SensorReading>,
    ) {
        // A filter command arrives, so enable forwarding for this key
        forwardingEnabled.update(true)

        // Calculate the timestamp when forwarding should be disabled
        val timerTimestamp = ctx.timerService().currentProcessingTime() + value.f1
        val curTimerTimestamp = disableTimer.value() ?: 0L

        // Check if the new timer should be later than any existing timer
        if (timerTimestamp > curTimerTimestamp) {
            // If there's an old timer, delete it
            if (curTimerTimestamp != 0L) {
                ctx.timerService().deleteProcessingTimeTimer(curTimerTimestamp)
            }
            // Register the new timer and store its timestamp in state
            ctx.timerService().registerProcessingTimeTimer(timerTimestamp)
            disableTimer.update(timerTimestamp)
        }
    }

    /** This method is called when a processing time timer fires for a key. */
    override fun onTimer(
        timestamp: Long,
        ctx: OnTimerContext,
        out: Collector<SensorReading>,
    ) {
        // The timer has fired, so we disable forwarding and clear the state for this key
        forwardingEnabled.clear()
        disableTimer.clear()
    }
}

/**
 * A factory object for creating filter switch sources.
 * Encapsulating this logic keeps the main job definition clean and focused.
 */
object FilterSwitch {
    /**
     * Creates a stream of Tuple2<String, Long> filter commands.
     * Elements are emitted with a 1-second delay between them to simulate a real-world scenario.
     */
    fun createFilterSwitchSource(env: StreamExecutionEnvironment): DataStreamSource<Tuple2<String, Long>> {
        val filterCommands =
            listOf(
                Tuple2("sensor_2", 5 * 1000L),
                Tuple2("sensor_7", 6 * 1000L),
                Tuple2("sensor_2", 10 * 1000L),
            )

        val generatorSource =
            DataGeneratorSource(
                { index -> filterCommands[index.toInt()] },
                filterCommands.size.toLong(),
                RateLimiterStrategy.perSecond(1.0),
                Types.TUPLE(Types.STRING, Types.LONG),
            )

        return env.fromSource(
            generatorSource,
            WatermarkStrategy.noWatermarks(),
            "Filter Switch Generator",
        )
    }
}
